### PR TITLE
add sam_mixin.py

### PR DESCRIPTION
**Pull request by jesicasusanto**
Issue : #168 
This PR adds sam_mixin.py with SAM model [https://github.com/facebookresearch/segment-anything](url) to get a bounding box of each mask.

steps to reproduce :
```
git remote add jesicasusanto https://github.com/jesicasusanto/PAT.git
git fetch jesicasusanto
git checkout feat/sam_mixin
pip install -r requirements.txt 
python -m puterbot.replay DemoReplayStrategy
```

then, wait to see the visualization of each segmented screenshot.
Here are the visualizations of `get_screenshot_bbox` :
image

Here are the visualizations of `get_click_event_bbox`

- before SAM :
- image

- after SAM :
- Screenshot (357)

- before SAM (unresized for better visualization) :
- image

- after SAM (unresized for better visualization) :
- Screenshot (349)

### CODE REVIEW

File: openadapt/strategies/mixins/sam_mixin.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            image_resized = resize_image(screenshot.image)
+            array_resized = np.array(image_resized)
+            self.sam_predictor.set_image(array_resized)
+            input_point = np.array([[screenshot.action_event.mouse_x, screenshot.action_event.mouse_y]])
+            input_label = np.array([1])
+            masks, scores, logits = self.sam_predictor.predict(
+            point_coords=input_point,
+            point_labels=input_label,
+            multimask_output=True,
+            )
```
Comment (abrichr) [inline openadapt/strategies/mixins/sam_mixin.py:103]: Please fix indentation with `black`

**Review by abrichr — CHANGES_REQUESTED** [review]
Thanks @jesicasusanto ! In this and all future PRs please include an example that the reader can run in order to verify that it works as intended. I believe your code will break due to the indentation problem but I can't easily verify this because there is no example usage.

File: openadapt/strategies/mixins/sam.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        logger.info(f"{len(screenshot.action_event)=}")
+        for action_event in screenshot.action_event:
+            if action_event.name in "click" and action_event.mouse_pressed == True:
+                logger.info(f"{action_event=}")
+                image_resized = resize_image(screenshot.image)
+                array_resized = np.array(image_resized)
+
+                # Resize mouse coordinates
+                resized_mouse_x = int(action_event.mouse_x * 0.1)
+                resized_mouse_y = int(action_event.mouse_y * 0.1)
```
Comment (abrichr) [inline openadapt/strategies/mixins/sam.py:102]: Please specify the RESIZE_RATIO at a constant at the top of this file

File: openadapt/strategies/mixins/sam.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+        """
+        image_resized = resize_image(screenshot.image)
+        array_resized = np.array(image_resized)
+        masks = self.sam_mask_generator.generate(array_resized)
+        bbox_list = []
+        for mask in masks:
+            bbox_list.append(mask["bbox"])
+            show_mask(mask["segmentation"], plt.gca())
+            show_box(mask["bbox"], plt.gca())
```
Comment (abrichr) [inline openadapt/strategies/mixins/sam.py:78]: Please add a SHOW_PLOTS constant at the top of this file to control whether these functions are called. Also add a parameter `show_plots=SHOW_PLOTS` to get_screenshot_bbox

File: openadapt/strategies/naive.py [hunk 4]
```diff
@@ -52,7 +52,7 @@ def get_next_action_event(
         if self.action_event_idx >= num_action_events:
             # TODO: refactor
             raise StopIteration()
-        action_event = action_events[self.action_event_idx]
+        action_event = action_events[self.action_event_idx]
```
Comment (abrichr) [inline openadapt/strategies/naive.py:55]: Please remove whitespace 🙏

Comment (jesicasusanto) [reply openadapt/strategies/naive.py:55]: done

**Comment by jesicasusanto** [comment]
Hi @abrichr, I have resolved all your comments. Moreover, I have found a bug when segmenting the click events with SAM model (explained more clearly in the above comment linked [here](https://github.com/MLDSAI/OpenAdapt/pull/169#issuecomment-1576794110)). I have also resolved this bug in this [commit](https://github.com/MLDSAI/OpenAdapt/pull/169/commits/90be2d1763789dcae8220504367a141c224cfe26). Please let me know if there is anything else I should be looking into. Thank you!

**Comment by jesicasusanto** [comment]
Hi @abrichr , 
30% Resolution : 1, 3, 4, 5
image
image
image
image

20% Resolution :
image
image
image

image

**Review by abrichr — APPROVED** [review]

**Comment by abrichr** [comment]
Implements https://github.com/MLDSAI/OpenAdapt/issues/15
